add dnd-character exercise

oxe-i opened the pull request:
I've followed the C implementation for the exercise, which doesn't check randomness of score generation. This also happens in the JS track and in the C++ track.

I've tried to implement a randomness check. After some research, it seemed that one commonly used test for that is a chi-squared test, which has a straightforward implementation. However, after extensive testing, I've noticed that once in a while the test simply failed even when using RDRAND (which theoretically has cryptographic level of randomness, by using hardware-generated entropy). I'm not sure why. 

Since this is outside my field of expertise and since many other tracks simply decided to not test for randomness, I've followed their lead.

Discussion on forum: https://forum.exercism.org/t/add-dnd-character-exercise/18981

oxe-i commented:
I'm not sure what the problem is. The run shows this error:

"The `exercises.foregone` array contains the slug `dnd-character`, but there is an implemented exercise with that slug:
/home/runner/work/x86-64-assembly/x86-64-assembly/config.json"

I've just checked, there's no dnd-character exercise in the x86-64-assembly track.

keiravillekode commented:
With regards to `grep` `parallel-letter-frequency` `bank-account` the online test runner uses Linux, so these could all be implemented using Linux system calls and pthreads. But we have students using Windows and Mac, working on their local machines, so I have kept calling the operating system out-of-scope for this track.

With regards to dnd-character, there is the question of testing "Is this random enough?" For example, if a student implements the dice roll to always return 3, do we want to accept that? Some tracks do, some don't. We could call the student's functions many times, and compare the resulting frequency distribution with what is expected. Unusual frequency distributions will occur occasionally but we might decide that something like a 0.01% false rejection rate is acceptable. There is also the question of the source of randomness. For example, do we pass in a seed? Overall, I concluded this exercise isn't something I wanted to implement or review for this track. If you and @kotp disagree with me, I won't be offended.

oxe-i commented:
Well, I understand your concerns regarding testing randomness. I've thought of them myself while implementing both this exercise and robot-name. I've tried to implement a check for distribution here on this one, but it didn't seem to work as well as I expected.

This problem seems to be common in other tracks. I've checked C, C++ and JS (languages I have some familiarity with) and they all implemented dnd-character without any test for randomness.

However, I tend to think that it's better to implement an imperfect exercise, which can then be eventually (and hopefully) improved, than not to implement it.

Besides, both exercises (robot-name and dnd-character) have other facets to them. Dnd-character requires using a struct with many elements, 7, which I believe is interesting enough for an easy exercise in x86-64-assembly. And robot-name requires having a map/dict, which is not trivial in x86-64-assembly.

oxe-i commented:
Regarding bank-account, I believe that right now the x86-64 track already locks windows users out, because it doesn't support Microsoft calling convention. And I believe (but I'm not 100% sure) that both macOS and Linux have native support to pthreads, it being just a matter of including the C header and setting the linking option in Make.

As for grep and parallel-letter-frequency, I tend to agree with you

kotp commented:
I agree with @oxe-i that the randomness test is not the killer feature here, it is the rest of it that is interesting in terms of mentoring.

Randomness testing is not a low hanging fruit, it is a tough problem solve, since any kind of sort is a sort, and due to the random nature, if we know it is random, then it makes sense that it is possible to even flip a coin and get 50 consecutive same answers.  So simply testing that you get something 'different' is not proof, just as getting something 'same' is not proof of it not being random.

If we want this exercise, it should be for the other reasons rather than the randomness.

I have no strong feelings one way or another, and this work has been done, and is referenced, and when the time is right, here for the continuation and bringing it in.  Now or later.  And I also do not mentor this track often (or at least not often enough) to have to bear the brunt of the discussion, when randomness comes up.

I mean to say that if we are not concerned about randomness and only in that we get an expected value, that is enough for me.  Our students are smart enough to recognize that they can improve on their code by working through the random or that it is OK not to.  Though whichever happens (sometimes false failures for testing random) or not testing robustly enough in the students mind, we should note it in the reading material for them.

oxe-i commented:
Ok then. Let me have another go into a randomness check. I'll probably need some time to study and prepare something. Once I get this done, or if I can't, I'll warn you people here and then we decide what to do next.

If we go along with the exercise, either with randomness check or not, instead of modifying the canonical instruction, I could add some append explaining how things are done in the track.

oxe-i commented:
Ok then. Let me have another go into a randomness check. I'll probably need some time to study and prepare something. Once I get this done, or if I can't, I'll warn you people here and then we decide what to do next.

If we go along with the exercise, either with randomness check or not, instead of modifying the canonical instruction, I could add some append explaining how things are done in the track.

oxe-i commented:
It seems I might have done something wrong in the implementation of the algorithm. I'll double check things here.

kotp commented:
Another way to think about the randomness is to make a comment (in analyzer or as automated feedback tooling) about "Feeling good if you have solved this in a way that gives random results for the attributes" and not worrying too much if they did not.  Then we can test that the results are only within the range of what we want, failing if they are doing the random thing when and if they end up with an "out of bounds" value.

I do not like that any passing test is a false positive, really, since the student may not run the test more than the number of times that they would have to in order to get a failing test (given the nature of "random").  But we have ways to communicate this... and ways that communicate it regardless of their solution when passing the tests, and without having to try to prove randomness.

kotp commented:
I think the work that you have done satisfies the "communication" part of the potential passing false positive.  (I realized that my prior comment might have suggested otherwise.)

kotp commented on exercises/practice/dnd-character/.docs/instructions.append.md:
```diff
@@ -0,0 +1,7 @@
+## x86-64 Assembly Specific Notes
+
+In the x86-64-assembly track, the randomness of the ability() function is checked by a [Chi-squared test](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test), considering a [p value](https://en.wikipedia.org/wiki/P-value) of 0.001.
```
Consider using reference links here, it makes the reading easier if reading in plain text, and means that all of our links, if needing to be maintained, are in one area.

keiravillekode replied:
> reference links

In case readers are unfamiliar with the term:

https://daringfireball.net/projects/markdown/syntax#link

https://exercism.org/docs/building/markdown/markdown

oxe-i replied:
Have done!

oxe-i commented:
> I do not like that any passing test is a false positive, really, since the student may not run the test more than the number of times that they would have to in order to get a failing test (given the nature of "random"). But we have ways to communicate this... and ways that communicate it regardless of their solution when passing the tests, and without having to try to prove randomness.

I'd rather not have this possibility. But I think that it's impossible to assert randomness in any definitive way, as you have already commented above.

Right now, I believe that this test I've implemented yield good results, since a correct (random) result failing is very unlikely (in theory, the odds would be of 0.1%) and every non-random algorithm I've tested (the most obvious ones) didn't pass. But both things (a correct result failing and a wrong result passing) are possible outcomes, unfortunately.

So, my question now would be: what do we prefer? Not testing randomness and just communicating that a random result would be ideal, or testing randomness and communicating the possibility of a false positive/negative?

Right now, both options seem feasible and it's just a matter of choosing what would be the best approach for this track.

To be honest, I believe most people don't want to cheat and communicating the objectives for the exercise should be enough. Testing randomness (or not) is mostly an extra, in my opinion.

oxe-i commented:
I've taken the "magic" number that corresponds to this percentage from a table. There's no lower p there, but the wikipedia entry mentions that this value can be calculated (it's not empirical). So I believe I can try to find a formula to get to 0.01%. 

Since it's late already here, tomorrow I'll do some research on this. One other possible improvement would be to use a multiple of 1296 (the number of different rolls of 4d6) for the number of iterations to get exact values for the expected table.

oxe-i commented:
@keiravillekode @kotp 

I've adjusted the test so that the chance of a correct (random) score failing is now 0.01%. I've also adjusted the number of iterations so that the expected table now holds exact values.

Those frequencies in the expected table were calculated numerically. If anyone wants to check, here's the code (in C++): https://onlinegdb.com/6VQe2aKaO

oxe-i commented on exercises/practice/dnd-character/.docs/instructions.append.md:
```diff
@@ -0,0 +1,10 @@
+## x86-64 Assembly Specific Notes
+
+In the x86-64-assembly track, the randomness of the ability() function is checked by a [Chi-squared test][chi-squared-test], considering a [p value][p-value] of 0.001.
```
Will do

oxe-i commented:
I've added the changes. Please, let me know if there's anything else. And thank you for your review!
